combat-trainer.lic

```diff
@@ -2894,18 +2900,35 @@ class AttackProcess
     echo("Time now: #{Time.now.to_i}") if $debug_mode_ct
     echo("@firing_timer: #{@firing_timer}") if $debug_mode_ct
     echo("@firing_delay: #{@firing_delay}") if $debug_mode_ct
-    case bput(command, 'isn\'t loaded', 'There is nothing', 'But your', /you (fire|poach|snipe) an? (?:.*\s)?(arrows?|bolt|shard|rock|sphere|clump|coral|fist|holder|lump|patella|pellet|pulzone|quadrello|quarrel|quill|stone|stopper|verretto|crumb|spine|mantrap spike|tiny dragon|icicle|fang|scale|grey-black spike|bacon strip|page|naga|thorn|fragment).* at/i, 'I could not find', 'with no effect and falls to the ground', 'Face what', 'How can you (poach|snipe)', 'you don\'t feel like fighting right now', 'That weapon must be in your right hand to fire')
+
+    # Occasionally the game borks and the phrases like "Face what?"
+    # appear before the text that you fired your weapon.
+    # So, we use a flag to see if the message ever appears
+    # rather than relying on the first message returned from bput.
+    Flags.add('ct-ranged-ammo', $CT_ATTACK_PROCESS_AMMO_REGEX)
+
+    result = bput(command, $CT_ATTACK_PROCESS_AMMO_REGEX, 'isn\'t loaded', 'There is nothing', 'But your', 'I could not find', 'with no effect and falls to the ground', 'Face what', 'How can you (poach|snipe)', 'you don\'t feel like fighting right now', 'That weapon must be in your right hand to fire', 'But you don\'t have a ranged weapon in your hand to fire with')
+    waitrt?
+
+    case result
     when /How can you (poach|snipe)/
       shoot_aimed('shoot', game_state)
-    when /you (fire|poach|snipe) an? (?:.*\s)?(arrows?|bolt|shard|rock|sphere|clump|coral|fist|holder|lump|patella|pellet|pulzone|quadrello|quarrel|quill|stone|stopper|verretto|blowgun dart|crumb|spine|mantrap spike|tiny dragon|icicle|fang|scale|grey-black spike|bacon strip|page|naga|thorn|fragment).* at/i
+    when /That weapon must be in your right hand to fire|But you don't have a ranged weapon in your hand to fire with/
+      fput('swap')
+      shoot_aimed(command, game_state)
+    when /you don't feel like fighting right now/
+      pause 1
+      shoot_aimed(command, game_state)
+    end
+
+    # If detected that you shot then stow the ammo.
+    if Flags['ct-ranged-ammo']
       game_state.action_taken
-      ammo = Regexp.last_match(2)
+      ammo = Flags['ct-ranged-ammo'][$CT_ATTACK_PROCESS_AMMO_REGEX_GROUP].to_s
```

Comment:
Instead of hardcoding an integer index we should consider making `events.lic` smarter and allow it to return matches against named match groups. We can do this with one small changes to events.

Store raw match data instead of the array.
```
    Flags.flags[key] = matches
```

```
irb(main):016:0> reg = /(?<name>\w+) .*/
=> /(?<name>\w+) .*/
irb(main):017:0> stuff = "Bob of the awesome".match(reg)
=> #<MatchData "Bob of the awesome" name:"Bob">
irb(main):018:0> stuff[1]
=> "Bob"
irb(main):019:0> stuff[:name]
=> "Bob"
```

This allows us seamlessly going forward to use `Flags['ct-ranged-ammo'][:noun]` (note the to_s is redundant here).

Reply:
Yes, thank you! I was bummed it only let me retrieve by index. I can look at updating events.lic per your suggestion.

Reply:
Update to `events.lic` submitted per https://github.com/rpherbig/dr-scripts/pull/4894